### PR TITLE
Source GitHub: Add option to pull commits from user-specified branches

### PR DESCRIPTION
**Pull request by cjwooo**
## What
Currently, the GitHub source only pulls commits from each repository's default branch. We want to pull commits from multiple branches.

## How
Add an option to specify a space-delimited list of repository branches to pull commits for. If no branches are specified for a repository, the source will pull only the default branch as it does currently. Cursor field state is tracked per branch instead of per repository.

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Comment by cjwooo** [comment]
I just realized this might break on reading existing Github state of previous versions. Any suggestions on how to resolve this?

File: airbyte-integrations/connectors/source-github/source_github/source.py [hunk 1]
```diff
@@ -102,18 +102,40 @@ def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) ->
     def streams(self, config: Mapping[str, Any]) -> List[Stream]:
         authenticator = self._get_authenticator(config["access_token"])
         repositories = self._generate_repositories(config=config, authenticator=authenticator)
-        organizations = list({org.split("/")[0] for org in repositories})
+        organizations = list(set({org.split("/")[0] for org in repositories}))
```
Comment (yevhenii-ldv) [inline airbyte-integrations/connectors/source-github/source_github/source.py:105]: There is no need for this. `{org.split("/")[0] for org in repositories}` already creates `set`, which filters out duplicate elements.

File: airbyte-integrations/connectors/source-github/source_github/source.py [hunk 2]
```diff
@@ -102,18 +102,40 @@ def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) ->
...
         full_refresh_args = {"authenticator": authenticator, "repositories": repositories}
         incremental_args = {**full_refresh_args, "start_date": config["start_date"]}
         organization_args = {"authenticator": authenticator, "organizations": organizations}
 
+        # Get the default branch for each repository
+        default_branches = {}
+        repository_stream = Repositories(authenticator=authenticator, organizations=organizations)
```
Comment (yevhenii-ldv) [inline airbyte-integrations/connectors/source-github/source_github/source.py:112]: The user repository may not belong to any organization, and in this case we will probably get an error.

Comment (yevhenii-ldv) [reply airbyte-integrations/connectors/source-github/source_github/source.py:112]: I think, as variant, in this case, we can expand the `RepositoryStats `stream with the default_branch field and use it, as variant.

File: airbyte-integrations/connectors/source-github/source_github/streams.py [hunk 3]
```diff
@@ -645,17 +645,91 @@ class Commits(IncrementalGithubStream):
...
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        repository = stream_slice["repository"]
+        branches_stream = Branches(authenticator=self.authenticator, repositories=[repository])
+        branches = [branch["name"] for branch in branches_stream.read_records(sync_mode=SyncMode.full_refresh, stream_slice=stream_slice)]
```
Comment (yevhenii-ldv) [inline airbyte-integrations/connectors/source-github/source_github/streams.py:722]: Why do we use this, if we have self.branches_to_pull?

**Review by yevhenii-ldv — CHANGES_REQUESTED** [review]
Please see comments

**Comment by yevhenii-ldv** [comment]
@cjwooo Regarding this [question](https://github.com/airbytehq/airbyte/pull/5931#issuecomment-915568964). 
I think that we can check which state the user is currently using, and in the output, in any case, we will receive a new state.
Something like this (Very draft version):
```
    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]):
        state_value = latest_cursor_value = latest_record.get(self.cursor_field)
        current_repository = latest_record["repository"]
        current_branch = latest_record["branch"]

        if current_stream_state.get(current_repository):
            repository_commits_state = current_stream_state[current_repository]
            if repository_commits_state.get(self.cursor_field):
                if  current_branch == repository["default_branch"]: # It is necessary to somehow throw `repository["default_branch"]` parameter here. 
                    state_value = max(latest_cursor_value, repository_commits_state [self.cursor_field])
                else:
                    state_value = latest_cursor_value
            elif repository_commits_state.get(current_branch, {}).get(self.cursor_field):
                state_value = max(latest_cursor_value, repository_commits_state[current_branch][self.cursor_field])
        if current_repository not in current_stream_state:
            current_stream_state[current_repository] = {}
        current_stream_state[current_repository][current_branch] = {self.cursor_field: state_value}
        return current_stream_state
```

File: airbyte-integrations/connectors/source-github/source_github/streams.py (hunk 3 above)
Comment (cjwooo) [reply airbyte-integrations/connectors/source-github/source_github/streams.py:722]: That was from my outdated code, will fix.

File: airbyte-integrations/connectors/source-github/source_github/source.py (hunk 2 above)
Comment (cjwooo) [reply airbyte-integrations/connectors/source-github/source_github/source.py:112]: Ah yes, the Repositories stream actually fails if a non-org repo is specified right now.

**Comment by cjwooo** [comment]
@yevhenii-ldv This PR is ready for review again.

**Comment by yevhenii-ldv** [comment]
Hey @cjwooo!
I looked at the code, LGTM, tomorrow (for my timezone), I will create a copy of this PR-and inside the Airbyte project (not in the fork branch) and run tests to make sure everything is fine. If everything works fine, tomorrow I will release a new version of the connector with your changes.

Thank you!

**Review by yevhenii-ldv — APPROVED** [review]

**Comment by marcosmarxm** [comment]
thanks a lot both! 🎉

**Comment by yevhenii-ldv** [comment]
Merged as #6223

**Pull request closed without merging**